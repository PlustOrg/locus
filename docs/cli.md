# Command-Line Interface (CLI)

The Locus CLI is the primary tool for managing your Locus projects. It provides a set of commands for common development tasks, from creating a new project to building and deploying it.

## `locus new <name>`

Scaffolds a new Locus project.

-   **Usage**: `locus new my-app`
-   **Description**: Creates a new directory with the given name and initializes it with a basic Locus project structure, including a `Locus.toml` configuration file.

## `locus build`

Builds the project.

-   **Usage**: `locus build`
-   **Description**: Compiles the `.locus` files into a full-stack web application. The generated code is written to the output directory, which is specified by the `--out` flag.
-   **Options**:
    -   `--src <dir>`: The source directory for `.locus` files. Defaults to the current directory.
    -   `--out <dir>`: The output directory for the generated code. Defaults to `generated`.

## `locus dev`

Starts the development server.

-   **Usage**: `locus dev`
-   **Description**: Starts a development server that watches for changes to your `.locus` files and automatically rebuilds the project.
-   **Options**:
    -   `--src <dir>`: The source directory for `.locus` files. Defaults to the current directory.

## `locus db <subcommand>`

Manages the database.

-   **Usage**: `locus db migrate|studio`
-   **Description**: Provides subcommands for managing the database:
    -   `migrate`: Runs database migrations.
    -   `studio`: Opens the Prisma Studio to view and manage your data.

## `locus deploy <env>`

Deploys the project.

-   **Usage**: `locus deploy production`
-   **Description**: Deploys the project to the specified environment. The deployment configuration is defined in `Locus.toml`.
