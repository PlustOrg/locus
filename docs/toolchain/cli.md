# Locus CLI Reference

The `locus` command-line interface is your main tool for managing Locus projects. It provides commands for project creation, development, building, testing, database management, deployment, and extensibility.

## Core Commands

### `locus new <project_name>`
Scaffold a new project with boilerplate files and configuration.

### `locus dev`
Start the development server with hot-reloading for frontend and backend.

#### Error Output Format

Both `locus build` and `locus dev` support structured error output for editor/tool integrations:

- `--errors pretty` (default): Human-friendly boxed messages with codeframes.
- `--errors json`: One JSON object per error written to stderr.

JSON schema (stable fields):

```
{
	"code": "parse_error" | "lex_error" | "validation_error" | "merge_error",
	"message": string,        // Friendly message, e.g., "Expected ':' but found 'String'"
	"rawMessage": string,     // Underlying message from the parser/lexer/throw site
	"filePath": string | null,
	"line": number | null,
	"column": number | null,
	"length": number | null,  // Length of the highlighted span
	"heading": string         // e.g., "Parse Error", "Validation Error"
}
```

Example:

```
$ locus build --errors json
{"code":"parse_error","message":"Expected ':' but found 'String'","rawMessage":"Expecting token of type --> Colon <-- but found --> 'String' <--","filePath":"src/db.locus","line":4,"column":12,"length":6,"heading":"Parse Error"}
```

### `locus build`
Compile and optimize your project for production.

#### Generated Outputs
Deterministic artifacts (relative to `--out`, default `generated/`):

- `prisma/schema.prisma` – Prisma data model
- `routes/**/*.ts` & `server.ts` – Express API scaffolding
- `react/pages/*.tsx` – Generated React pages
- `react/components/*.tsx` – Generated React components
- `theme.css` – Design tokens as CSS custom properties
- `next-app/app/**/*` – Minimal Next.js App Router shell
- `next-app/public/theme.css` – Theme duplication for static serving
- `package.json` – Created only if absent (app name from `Locus.toml` `[app].name`)
- `README.md` – Quickstart (only if absent)

All files begin with:
```ts
// AUTO-GENERATED by Locus. DO NOT EDIT.
```

#### Flags
- `--out <dir>`: Output directory (default `generated`)
- `--errors <pretty|json>`: Error rendering style
- `--prisma-generate`: Run `npx prisma generate` after writing outputs
- `--dry-run`: Print planned files; perform no writes or side effects

Example dry run:
```bash
locus build --dry-run
[locus][build][dry-run] files that would be written:
 - prisma/schema.prisma
 - routes/user.ts
 - react/pages/Home.tsx
 ...
```

#### Determinism & Ordering
Artifacts are sorted (alphabetically or by entity/page name) to ensure stable diffs and snapshot-friendly output.

#### Side Effects
No dependency install, server start, or Prisma client generation unless explicitly requested with `--prisma-generate`.

#### Package Name Resolution
If `Locus.toml` contains:
```toml
[app]
name = "my-app"
```
the generated `package.json` uses `my-app`; otherwise falls back to `locus-generated-app`.

### `locus dev`
Start the development server with hot-reloading for frontend and backend.

#### Incremental Generation
Performs an initial full build (same artifacts as `locus build`), then regenerates only the files whose content changes. A content cache avoids redundant disk writes.

Change feedback:
- Summary: `[locus][dev] regenerated N files`
- Verbose list (when `LOCUS_DEBUG=1`): `[locus][dev][changed] react/pages/Home.tsx, prisma/schema.prisma, theme.css`

`package.json` & `README.md` are generated once if absent; never overwritten.

#### Shared Artifact Module
Both build and dev use a shared internal artifact assembler ensuring feature parity and single-point evolution (e.g., adding an OpenAPI generator in the future).

### `locus db migrate '<name>'`
Generate and apply database migrations based on your data model.

### `locus db studio`
Launch Prisma Studio to view and edit your database.

### `locus test`
Run all tests defined in your `.locus` files.

### `locus deploy`
Build and deploy your app to modern hosting platforms (Vercel, Railway, Supabase, etc.).

### `locus add <package>`
Install npm packages into the generated frontend or backend.

## Advanced Usage

- All commands work with zero configuration, but can be customized via `Locus.toml`.
- The CLI supports plugins for new commands and integrations.

See [Development Workflow](./development-workflow.md) and [Deployment](./deployment.md) for more details.

## Generated Package Manifest (Example)
```json
{
	"name": "my-app",
	"private": true,
	"scripts": {
		"dev:api": "node server.ts",
		"dev:next": "next dev next-app",
		"dev": "npm run dev:api & npm run dev:next",
		"prisma:generate": "prisma generate",
		"build": "next build next-app",
		"start": "next start next-app"
	},
	"dependencies": {
		"express": "^4.19.2",
		"@prisma/client": "^5.15.0",
		"cors": "^2.8.5",
		"react": "^18.3.1",
		"react-dom": "^18.3.1",
		"next": "^14.2.0"
	},
	"devDependencies": { "prisma": "^5.15.0" }
}
```
React / Next dependencies are only added if pages exist.

## Theming
`theme.css` (design tokens) is also copied to `next-app/public/theme.css` so it is served at `/theme.css` for the Next shell.

## Prisma Client Options
Generate the Prisma client via either:
1. `locus build --prisma-generate`
2. `npm run prisma:generate` inside the generated directory

Opt-in keeps builds fast and deterministic for CI and dry runs.

## Dry Run Scenarios
Use `--dry-run` for CI planning, editor previews, plugin development, or diff inspection without touching the filesystem.

## Incremental Strategy Details
Dev mode stores last written contents and rewrites only when changed, minimizing I/O churn and reducing risk of watcher loops.

## Roadmap (Planned Hooks)
- `--openapi` to emit OpenAPI specs
- `--emit-client-only` to skip server/UI emission
- Plugin API to register custom artifact contributors

Feedback or proposals welcome—open an issue or submit a design doc PR.

## Troubleshooting

Prisma client not found:
- Run `locus build --prisma-generate` or inside generated dir run `npm run prisma:generate`.
- Ensure `.env` has a valid `DATABASE_URL`.
## Dev Mode

### Startup Banner

Running `locus dev` now prints a concise banner summarizing the running environment:

```
┌────────────────────────────────────────────┐
│ App: my-app                                │
│ API:  http://localhost:3001  (routes: 3)   │
│ Web:  http://localhost:3000                │
│ Theme: ✓   Prisma: ✗ (run prisma generate) │
│ Watching: **/*.locus                       │
│ CORS: off  NODE_ENV: development           │
│ Ctrl+C to stop                             │
└────────────────────────────────────────────┘
```

Legend:
- API/Web: Base URLs. API port precedence: `API_PORT` > `PORT` > 3001.
- routes: Count of generated Express route files.
- Theme: Whether `theme.css` was generated.
- Prisma: Whether `@prisma/client` is resolvable (hint shown if missing).
- Watching: Glob pattern for incremental rebuild.
- CORS: Enabled when `ENABLE_CORS=1`.

The banner appears once after the initial build. Subsequent rebuilds log concise `[locus][dev] regenerated N files` messages. Use `LOCUS_DEBUG=1` for extra timing details (future enhancement).
`--quiet` suppresses the banner and startup lines (only regeneration or error lines remain).

With `LOCUS_DEBUG=1` each change batch prints an additional line:
```
[locus][dev][timing] batch=2 total=15 dt=42ms
```
Meaning: last batch rewrote 2 files, cumulative rewrites 15, 42ms since previous batch.

Port conflicts:
- API defaults to `PORT` env var or 3001. Set a free port: `PORT=4001 locus dev`.
- Next.js dev server (if present) uses 3000; set `PORT=4002` if you have another service at 3001.

Missing theme.css in browser:
- Confirm `theme.css` exists in output root and `next-app/public/theme.css`.
- Ensure landing page `<head>` loads `globals.css` which imports `/theme.css`.

CORS blocked requests:
- Set `ENABLE_CORS=1` when running dev/build server to enable basic CORS middleware.

Regeneration not triggering:
- Check that file is within watched src directory passed to `--src`.
- Use `LOCUS_DEBUG=1 locus dev` to print changed file list.

## Route Naming Policy

Generated Next routes use kebab-case derived from Locus page names (e.g., `UserProfile` -> `/user-profile`).

API entity routes default to the lowerCamel entity name (e.g., `User` -> `/user`). An experimental pluralization option (internal flag) transforms route mounts to plural forms (`/users`, `Category` -> `/categories`). This will surface as a formal CLI/config option in a future release.

Port selection precedence for the API server: `API_PORT` > `PORT` > 3001.

See also: [Common Errors & Fixes](../guides/common-errors.md).
