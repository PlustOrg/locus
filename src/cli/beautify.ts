import chalk from 'chalk';
import boxen from 'boxen';

const brandColor = chalk.hex('#00aaff'); // A nice blue for Locus brand

export function info(message: string) {
  console.log(brandColor('i ') + message);
}

export function success(message: string) {
  console.log(chalk.green('✓ ') + message);
}

export function warn(message: string) {
  console.log(chalk.yellow('! ') + message);
}

export function error(message: string) {
  console.log(chalk.red('✗ ') + message);
}

export function step(message: string) {
  console.log(chalk.bold.underline(`\nStep: ${message}\n`));
}

export function box(content: string, title?: string) {
    console.log(
        boxen(content, {
            padding: 1,
            margin: 1,
            borderStyle: 'round',
            borderColor: 'cyan',
            title: title ? brandColor.bold(title) : undefined,
            titleAlignment: 'center',
        })
    );
}

export function devServerBanner(info: {
  appName: string;
  apiPort: number;
  hasPages: boolean;
  nextPort?: number;
  theme: boolean;
  prismaClient: boolean;
  enableCors: boolean;
  watchPattern: string;
  routeCount: number;
  prismaHint: boolean;
}) {
  const lines: string[] = [];
  lines.push(`${chalk.bold('App')}: ${chalk.cyan(info.appName)}`);
  lines.push(`${chalk.bold('API')}: ${chalk.cyan(`http://localhost:${info.apiPort}`)} (routes: ${info.routeCount})`);
  if (info.hasPages) lines.push(`${chalk.bold('Web')}: ${chalk.cyan(`http://localhost:${info.nextPort || 3000}`)}`);
  lines.push('');
  lines.push(`${chalk.bold('Theme')}: ${info.theme ? chalk.green('✓') : chalk.red('✗')}   ${chalk.bold('Prisma')}: ${info.prismaClient ? chalk.green('✓') : chalk.red('✗')}${info.prismaHint ? chalk.yellow(' (run prisma generate)') : ''}`);
  lines.push(`${chalk.bold('Watching')}: ${chalk.dim(info.watchPattern)}`);
  lines.push(`${chalk.bold('CORS')}: ${info.enableCors ? chalk.green('on') : chalk.red('off')}  ${chalk.bold('NODE_ENV')}: ${chalk.dim(process.env.NODE_ENV || 'development')}`);
  lines.push('');
  lines.push(chalk.dim('Ctrl+C to stop'));

  box(lines.join('\n'), 'Locus Dev Server');
}

export function finalMessage(appName: string) {
    const message = `
${brandColor.bold('Congratulations! Your Locus application is ready.')}

Next steps:
  1. ${chalk.cyan('cd generated')}
  2. ${chalk.cyan('npm install')}
  3. ${chalk.cyan('npm run dev')}

Happy coding!
`;
    box(message, `Locus App: ${appName}`);
}
