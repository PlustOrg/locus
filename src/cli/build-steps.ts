import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { spawnSync } from 'child_process';
import chalk from 'chalk';
import { LocusError, errorToDiagnostic, Diagnostic } from '../errors';
import { parseLocus } from '../parser';
import { mergeAsts } from '../parser/merger';
import { validateUnifiedAst } from '../validator/validate';
import { PluginManager } from '../plugins/manager';
import {
  buildOutputArtifacts,
  buildPackageJson,
  buildGeneratedReadme,
  getAppName,
  buildTsConfig,
} from '../generator/outputs';
import { generateExpressApi, AuthConfig } from '../generator/express';
import { pLimit } from './utils/concurrency';
import { safeMkdir, safeWrite } from './utils/files';

/**
 * Parses all .locus files in the project.
 * @param files The list of .locus files to parse.
 * @param pluginMgr The plugin manager instance.
 * @returns An object containing the parsed ASTs, a map of file contents, and any diagnostics.
 */
export async function parseLocusFiles(
  files: string[],
  pluginMgr: PluginManager
): Promise<{
  asts: any[];
  fileMap: Map<string, string>;
  diagnostics: Diagnostic[];
  filePaths: string[];
}> {
  const asts: any[] = [];
  const fileMap = new Map<string, string>();
  const diagnostics: Diagnostic[] = [];
  const filePaths: string[] = [];

  for (const filePath of files) {
    try {
      const content = readFileSync(filePath, 'utf8');
      fileMap.set(filePath, content);

      try {
        await pluginMgr.onParseStart(filePath, content);
      } catch {
        // Ignore plugin errors during this hook
      }

      const ast = parseLocus(content, filePath);
      asts.push(ast);
      filePaths.push(filePath);
    } catch (e) {
      if (e instanceof LocusError || (e && (e as any).code)) {
        diagnostics.push(errorToDiagnostic(e as any));
        continue; // Continue parsing other files
      }
      throw new Error(`Failed to parse ${filePath}: ${(e as any)?.message || e}`);
    }
  }

  return { asts, fileMap, diagnostics, filePaths };
}

/**
 * Merges and validates the ASTs.
 * @param asts The array of ASTs to merge.
 * @param pluginMgr The plugin manager instance.
 * @returns The merged and validated AST.
 */
export async function mergeAndValidateAsts(
  asts: any[],
  filePaths: string[],
  pluginMgr: PluginManager
): Promise<any> {
  // Allow plugins to run file-parsed hooks
  for (let i = 0; i < asts.length; i++) {
    const ast = asts[i];
    const filePath = filePaths[i];
    try {
      await pluginMgr.onFileParsed(filePath, ast);
    } catch {
      // Collect errors inside the manager
    }
  }

  // Allow plugins to inject virtual ASTs
  try {
    await pluginMgr.onParseComplete(asts);
  } catch {
    // Ignore plugin errors
  }

  const allAsts = asts.concat(pluginMgr.virtualAsts);
  const mergedAst = mergeAsts(allAsts);

  // Allow plugins to validate the unified AST
  await pluginMgr.onValidate(mergedAst);
  validateUnifiedAst(mergedAst);

  return mergedAst;
}

/**
 * Generates and writes the output artifacts.
 * @param mergedAst The unified AST.
 * @param config The project configuration.
 * @param pluginMgr The plugin manager.
 * @param srcDir The source directory.
 * @param outDir The output directory.
 * @param opts Additional build options.
 * @returns A promise that resolves with the generation metadata.
 */
export async function generateAndWriteArtifacts(
  mergedAst: any,
  config: any,
  pluginMgr: PluginManager,
  srcDir: string,
  outDir: string,
  opts: {
    dryRun?: boolean;
    suppressWarnings?: boolean;
    emitJs?: boolean;
  }
): Promise<any> {
  await pluginMgr.onBeforeGenerate(mergedAst);

  // Detect auth configuration
  let auth: AuthConfig | undefined;
  if (config.auth?.jwtSecret || config.raw._sections?.auth) {
    const authSection = config.raw._sections?.auth || {};
    auth = {
      jwtSecret: config.auth?.jwtSecret,
      adapterPath: authSection.adapter,
      requireAuth: !!authSection.requireAuth,
    } as any;
    if (auth?.jwtSecret && !process.env.LOCUS_JWT_SECRET) {
      process.env.LOCUS_JWT_SECRET = auth.jwtSecret;
    }
  }

  const { files: artifacts, meta } = buildOutputArtifacts(mergedAst, { srcDir });

  if (config.warnings?.length) {
    meta.warnings = [...(meta.warnings || []), ...config.warnings];
  }

  // Augment express server if auth is configured
  if (auth) {
    const guardedPages = (mergedAst.pages || [])
      .filter((p: any) => p.guard)
      .map((p: any) => ({ name: p.name, role: p.guard.role }));
    const expressFiles = generateExpressApi(mergedAst.database?.entities || [], {
      auth,
      pagesWithGuards: guardedPages,
    });
    Object.assign(artifacts, expressFiles);
  }

  // Allow plugins to modify artifacts
  await pluginMgr.onAfterGenerate({ artifacts, meta });
  pluginMgr.runCustomGenerators(mergedAst);
  Object.assign(artifacts, pluginMgr.extraArtifacts);

  if (opts.suppressWarnings && meta.warnings?.length) {
    delete (artifacts as any)['GENERATED_WARNINGS.txt'];
  }

  meta.warnings = [...(meta.warnings || []), ...pluginMgr.warnings];
  (meta as any).pluginTimings = pluginMgr.timings;

  if (opts.dryRun) {
    const fileList = Object.keys(artifacts).sort();
    process.stdout.write(
      '[locus][build][dry-run] files that would be written:\n' +
        fileList.map((f) => ' - ' + f).join('\n') +
        '\n'
    );
    return {
      dryRun: true,
      filesPlanned: fileList,
      meta: { hasPages: meta.hasPages, warnings: meta.warnings },
    };
  }

  const sortedArtifacts = Object.entries(artifacts).sort(([a], [b]) =>
    a.localeCompare(b)
  );

  const limit = pLimit(6);
  await Promise.all(
    sortedArtifacts.map(([relativePath, content]) =>
      limit(async () => {
        const fullPath = join(outDir, relativePath);
        const dir = dirname(fullPath);
        await safeMkdir(dir);
        await safeWrite(fullPath, content as string);
      })
    )
  );

  // Ensure project scaffolding files exist
  const appName = getAppName(srcDir);
  const pkgJsonPath = join(outDir, 'package.json');
  if (!existsSync(pkgJsonPath)) {
    writeFileSync(pkgJsonPath, buildPackageJson(meta.hasPages, appName));
  }
  const readmePath = join(outDir, 'README.md');
  if (!existsSync(readmePath)) {
    writeFileSync(readmePath, buildGeneratedReadme());
  }
  const tsconfigPath = join(outDir, 'tsconfig.json');
  if (!existsSync(tsconfigPath)) {
    writeFileSync(tsconfigPath, buildTsConfig());
  }

  // Optionally compile TypeScript to JavaScript
  if (opts.emitJs) {
    try {
      const res = spawnSync(
        'npx',
        [
          'tsc',
          '--project',
          tsconfigPath,
          '--outDir',
          'dist',
          '--declaration',
          'false',
          '--emitDeclarationOnly',
          'false',
        ],
        { cwd: outDir, stdio: 'ignore' }
      );
      if (res.status !== 0) {
        process.stderr.write(
          '[locus][build] tsc exited with code ' + res.status + '\n'
        );
      }
    } catch {
      // Ignore compilation errors
    }
  }

  if (!opts.suppressWarnings && meta.warnings && meta.warnings.length) {
    for (const warning of meta.warnings) {
      process.stdout.write(chalk.yellow('[locus][warn] ' + warning + '\n'));
    }
  }

  return meta;
}
