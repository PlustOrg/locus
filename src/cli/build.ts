import { readdirSync, writeFileSync, mkdirSync, existsSync, readFileSync, statSync } from 'fs';
import { spawnSync } from 'child_process';
import { promises as fsp } from 'fs';
import { join, dirname } from 'path';
import { parseLocus } from '../parser';
import { mergeAsts } from '../parser/merger';
import { validateUnifiedAst } from '../validator/validate';
import { generatePrismaSchema } from '../generator/prisma';
import { generateExpressApi } from '../generator/express';
import { BuildError, GeneratorError, LocusError } from '../errors';
import { generateReactComponent, generateReactPage } from '../generator/react';
import { generateCssVariables } from '../generator/theme';
import { generateNextApp } from '../generator/next';
import { reportError, ErrorOutputFormat } from './reporter';
export async function buildProject(opts: { srcDir: string; outDir?: string; debug?: boolean; errorFormat?: ErrorOutputFormat; prismaGenerate?: boolean }) {
  const srcDir = opts.srcDir;
  const outDir = opts.outDir || join(srcDir, 'generated');
  const debug = !!opts.debug;
  const t0 = Date.now();

  let files: string[];
  try {
    files = findLocusFiles(srcDir);
  } catch (e) {
    throw new BuildError(`Failed to read source directory: ${srcDir}`, e);
  }
  const fileMap = new Map<string, string>();
  const tParse0 = Date.now();
  const asts = files.map(fp => {
    try {
      const content = typeof readFileSync === 'function' ? readFileSync(fp, 'utf8') : String(fp);
      fileMap.set(fp, content);
      return parseLocus(content as any, fp);
    } catch (e) {
      if (e instanceof LocusError || (e && (e as any).code)) {
  reportError((e as any) as LocusError, fileMap, opts.errorFormat);
        process.exit(1);
      }
      throw new BuildError(`Failed to parse ${fp}: ${(e as any)?.message || e}`, e);
    }
  });
  const tParse1 = Date.now();
  let merged;
  try {
    merged = mergeAsts(asts);
  } catch (e) {
    if (e instanceof LocusError || (e && (e as any).code)) {
      reportError((e as any) as LocusError, fileMap, opts.errorFormat);
      process.exit(1);
    }
    throw new BuildError(`Failed to merge ASTs: ${(e as any)?.message || e}`, e);
  }
  // Validate unified AST
  try {
    validateUnifiedAst(merged);
  } catch (e) {
    if (e instanceof LocusError || (e && (e as any).code)) {
  reportError((e as any) as LocusError, fileMap, opts.errorFormat);
      process.exit(1);
    }
    throw e;
  }
  const tMerge1 = Date.now();

  // helper to add generated header
  const withHeader = (content: string, kind?: string) => {
    const header = '// AUTO-GENERATED by Locus. DO NOT EDIT.' + (kind ? ` (${kind})` : '') + '\n';
    return content.startsWith('// AUTO-GENERATED') ? content : header + content;
  };

  // Prisma
  try {
    const schema = withHeader(generatePrismaSchema(merged.database), 'prisma schema');
    const prismaDir = join(outDir, 'prisma');
    if (!existsSync(prismaDir)) mkdirSync(prismaDir, { recursive: true });
    writeFileSync(join(prismaDir, 'schema.prisma'), schema);
  } catch (e) {
    if (e instanceof LocusError || (e && (e as any).code)) {
      reportError((e as any) as LocusError, fileMap, opts.errorFormat);
      process.exit(1);
    }
    throw new GeneratorError('Failed generating Prisma schema', e);
  }

  // Express
  try {
    const routes = generateExpressApi(merged.database.entities as any);
    const entries = Object.entries(routes).sort(([a], [b]) => a.localeCompare(b));
    const limit = pLimit(4);
    await Promise.all(entries.map(([p, c]) => limit(async () => {
      const full = join(outDir, p);
      const dir = dirname(full);
      await safeMkdir(dir);
      await safeWrite(full, withHeader(c, 'express route'));
    })));
    const serverPath = join(outDir, 'server.ts');
    if (existsSync(serverPath)) {
      const current = readFileSync(serverPath, 'utf8');
      if (!current.startsWith('// AUTO-GENERATED')) writeFileSync(serverPath, withHeader(current, 'express server'));
    }
  } catch (e) {
    if (e instanceof LocusError || (e && (e as any).code)) {
      reportError((e as any) as LocusError, fileMap);
      process.exit(1);
    }
    throw new GeneratorError('Failed generating Express API', e);
  }

  // React (pages & components)
  const reactBase = join(outDir, 'react');
  const pagesDir = join(reactBase, 'pages');
  const compsDir = join(reactBase, 'components');
  if (!existsSync(pagesDir)) mkdirSync(pagesDir, { recursive: true });
  if (!existsSync(compsDir)) mkdirSync(compsDir, { recursive: true });
  // ensure stable output ordering
  const sortedPages = [...(merged.pages as any[])].sort((a, b) => a.name.localeCompare(b.name));
  {
  const limit = pLimit(4);
  await Promise.all(sortedPages.map(p => limit(async () => {
      try {
  const code = withHeader(generateReactPage(p), 'react page');
  await safeWrite(join(pagesDir, `${p.name}.tsx`), code);
      } catch (e) {
        if (e instanceof LocusError || (e && (e as any).code)) {
          reportError((e as any) as LocusError, fileMap, opts.errorFormat);
          process.exit(1);
        }
        throw new GeneratorError(`Failed generating React page '${p.name}'`, e);
      }
    })));
  }

  // Theme CSS from design_system
  try {
    const css = withHeader(generateCssVariables(merged.designSystem), 'theme css');
    // Write to outDir root (Next app will import from /theme.css)
    writeFileSync(join(outDir, 'theme.css'), css);
  } catch {/* optional */}

  // Next.js minimal app scaffolding (app/ directory routing)
  try {
    const nextFiles = generateNextApp(merged.pages as any);
    for (const [rel, content] of Object.entries(nextFiles)) {
      const full = join(outDir, rel);
      const dir = dirname(full);
      if (!existsSync(dir)) mkdirSync(dir, { recursive: true });
      writeFileSync(full, withHeader(content, 'next'));
    }
    // copy theme.css into next-app/public
    const themePath = join(outDir, 'theme.css');
    if (existsSync(themePath)) {
      const pub = join(outDir, 'next-app', 'public');
      if (!existsSync(pub)) mkdirSync(pub, { recursive: true });
      writeFileSync(join(pub, 'theme.css'), readFileSync(themePath, 'utf8'));
    }
  } catch {/* optional */}
  const sortedComps = [...(merged.components as any[])].sort((a, b) => a.name.localeCompare(b.name));
  {
  const limit = pLimit(4);
  await Promise.all(sortedComps.map(c => limit(async () => {
      try {
  const code = withHeader(generateReactComponent(c), 'react component');
  await safeWrite(join(compsDir, `${c.name}.tsx`), code);
      } catch (e) {
        if (e instanceof LocusError || (e && (e as any).code)) {
          reportError((e as any) as LocusError, fileMap, opts.errorFormat);
          process.exit(1);
        }
        throw new GeneratorError(`Failed generating React component '${c.name}'`, e);
      }
    })));
  }

  if (debug) {
    const t1 = Date.now();
    const timings = {
      files: files.length,
      parseMs: tParse1 - tParse0,
      mergeMs: tMerge1 - tParse1,
      generateMs: t1 - tMerge1,
      totalMs: t1 - t0,
    };
  process.stdout.write('[locus][build][timings] ' + JSON.stringify(timings) + '\n');
  }

  // package.json & README
  try {
    const pkgFile = join(outDir, 'package.json');
    if (!existsSync(pkgFile)) {
      const hasPages = (merged.pages as any[]).length > 0;
      const pkg = {
        name: 'locus-generated-app',
        private: true,
        version: '0.0.0',
        scripts: {
          'dev:api': 'node server.ts',
          ...(hasPages ? { 'dev:next': 'next dev next-app' } : {}),
          dev: hasPages ? 'npm run dev:api & npm run dev:next' : 'npm run dev:api',
          'prisma:generate': 'prisma generate',
          build: hasPages ? 'next build next-app' : 'echo "No Next app"',
          start: hasPages ? 'next start next-app' : 'node server.ts'
        },
        dependencies: {
          express: '^4.19.2',
          '@prisma/client': '^5.15.0',
          cors: '^2.8.5',
          ...(hasPages ? { react: '^18.3.1', 'react-dom': '^18.3.1', next: '^14.2.0' } : {})
        },
        devDependencies: { prisma: '^5.15.0' }
      } as any;
      writeFileSync(pkgFile, JSON.stringify(pkg, null, 2));
    }
    const readmeFile = join(outDir, 'README.md');
    if (!existsSync(readmeFile)) {
      const readme = `# Locus Generated App\n\nInstall deps:\n\n    npm install\n\nGenerate Prisma client (once):\n\n    npm run prisma:generate\n\nRun dev (API only or full if Next pages exist):\n\n    npm run dev\n\nBuild (Next pages):\n\n    npm run build\n\nThis directory is auto-generated; changes may be overwritten.\n`;
      writeFileSync(readmeFile, readme);
    }
  } catch {/* ignore */}

  if (opts.prismaGenerate) {
    try {
      const schema = join(outDir, 'prisma', 'schema.prisma');
      if (existsSync(schema)) {
        spawnSync('npx', ['prisma', 'generate', '--schema', schema], { stdio: 'ignore' });
      }
    } catch {/* ignore */}
  }

  return { outDir };
}

function findLocusFiles(dir: string): string[] {
  let entries: any;
  try {
    entries = readdirSync(dir, { withFileTypes: true } as any);
  } catch {
    entries = readdirSync(dir);
  }
  // If mocked fs returns string[]
  if (Array.isArray(entries) && typeof entries[0] === 'string') {
    return (entries as string[])
      .map(name => join(dir, name))
      .filter(p => p.endsWith('.locus'));
  }
  // Real dirents path
  const results: string[] = [];
  for (const entry of entries) {
    const full = join(dir, entry.name);
    if (entry.isDirectory && entry.isDirectory()) {
      results.push(...findLocusFiles(full));
    } else if (entry.isFile && entry.isFile() && full.endsWith('.locus')) {
      results.push(full);
    } else if (!entry.isDirectory && !entry.isFile) {
      // Fallback for environments lacking isFile/isDirectory on dirent
      try {
        const st = statSync(full);
        if (st.isDirectory()) results.push(...findLocusFiles(full));
        else if (st.isFile() && full.endsWith('.locus')) results.push(full);
      } catch {
        // ignore
      }
    }
  }
  return results;
}

function pLimit(concurrency: number) {
  let active = 0;
  const queue: Array<() => void> = [];
  const next = () => {
    if (active >= concurrency) return;
    const job = queue.shift();
    if (!job) return;
    active++;
    job();
  };
  return function <T>(fn: () => Promise<T>): Promise<T> {
    return new Promise<T>((resolve, reject) => {
      const run = () => {
        Promise.resolve(fn()).then(
          (v) => { active--; resolve(v); next(); },
          (e) => { active--; reject(e); next(); }
        );
      };
      queue.push(run);
      next();
    });
  };
}

async function safeMkdir(dir: string) {
  try {
    if ((fsp as any)?.mkdir) return await (fsp as any).mkdir(dir, { recursive: true });
  } catch {/* ignore */}
  try { mkdirSync(dir, { recursive: true }); } catch {/* ignore */}
}

async function safeWrite(path: string, content: string) {
  try {
    if ((fsp as any)?.writeFile) return await (fsp as any).writeFile(path, content, 'utf8');
  } catch {/* fall through to sync */}
  writeFileSync(path, content);
}
