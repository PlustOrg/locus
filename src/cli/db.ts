import { execAsync } from './utils';

/**
 * Run Prisma database migration.
 */
export async function runDbMigrate(_opts: { cwd?: string }) {
  await execAsync('npx prisma migrate dev');
}

/**
 * Open Prisma Studio GUI for database management.
 */
export async function runDbStudio(_opts: { cwd?: string }) {
  await execAsync('npx prisma studio');
}
