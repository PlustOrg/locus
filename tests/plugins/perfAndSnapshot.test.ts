import { mkdtempSync, writeFileSync, rmSync, readFileSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';
import { buildProject } from '../../src/cli/build';

function tdir(){ return mkdtempSync(join(tmpdir(), 'locus-plugin-perf-')); }

describe('Plugin performance guard & snapshot', () => {
  test('slow hook triggers performance warning', async () => {
    const dir = tdir();
    try {
      writeFileSync(join(dir,'app.locus'),'component P { ui { <div/> } }');
      writeFileSync(join(dir,'locus.plugins.js'), `module.exports=[{name:'slowPerf', apiVersion:1, onAfterGenerate(){ const start=Date.now(); while(Date.now()-start<30){} }}];`);
      process.env.LOCUS_PLUGIN_HOOK_WARN_MS = '10';
      const out = join(dir,'generated');
      const res = await buildProject({ srcDir: dir, outDir: out });
  expect(res.meta.warnings.some((_w:string)=>/performance/i)).toBe(true);
    } finally { delete process.env.LOCUS_PLUGIN_HOOK_WARN_MS; rmSync(dir,{recursive:true,force:true}); }
  });

  test('custom generator output snapshot', async () => {
    const dir = tdir();
    try {
      writeFileSync(join(dir,'app.locus'),'component S { ui { <div/> } }');
      writeFileSync(join(dir,'locus.plugins.js'), `module.exports=[{name:'snapGen', apiVersion:1, onBeforeGenerate(u,ctx){ ctx.registerGenerator('snap',(u)=>({'extra/hello.txt':'hello-world'})); }}];`);
      const out = join(dir,'generated');
      await buildProject({ srcDir: dir, outDir: out });
      const content = readFileSync(join(out,'extra/hello.txt'),'utf8');
      expect(content).toMatchSnapshot('plugin-extra-hello');
    } finally { rmSync(dir,{recursive:true,force:true}); }
  });
});
