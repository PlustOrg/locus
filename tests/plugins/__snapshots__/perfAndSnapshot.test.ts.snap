// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin performance guard & snapshot custom generator output snapshot plugin-extra-hello 1`] = `"hello-world"`;

exports[`Plugin performance guard & snapshot custom generator output snapshot: plugin-extra-hello 1`] = `"hello-world"`;
